File: C/array/queue.c
/**
 * Author: Axel Ariel Saravia
 */

/*
    IMPORTANT

    the T type is the signed type

    queue struct is {
        T* content;
        size_t capacity;
        size_t length;
        size_t head;
    }
*/

# include "queue.h"

queue queue_create() {
    T* content = calloc(8, sizeof(T));
    return (
        !content
        ? (queue){0}
        : (queue) {
            .content = content,
            .capacity = 8,
            .head = 0,
            .length = 0,
        }
    );
}

size_t queue_size(queue const q) {
    return q.length;
}

T queue_get(queue const q) {
    return q.content[q.head];
}

queue queue_add(queue q, T const x) {
    if (q.length + 1 > q.capacity) {
        q = queue_resize(q);
    }
    size_t index = (q.head + q.length) % q.capacity;
    q.content[index] = x;
    q.length += 1;
    return q;
}

queue queue_remove(queue q) {
    T x = q.content[q.head];
    q.head = (q.head + 1) % q.capacity;
    q.length -= 1;

    if (8 < q.capacity && 3 * q.length <= q.capacity) {
        queue_resize(q);
    }

    return q;
}


queue queue_resize(queue q) {
    size_t const length = q.length;
    size_t const new_capacity = (
        8 < 2 * length
        ? 2 * length
        : 8
    );
    // this is never true when resize is called from add or remove methods
    if (new_capacity < length) {
        return q;
    }
    T* new_content = 0;
    size_t const old_capacity = q.capacity;
    size_t old_head = q.head;
    size_t new_head = old_head;
    if (old_capacity < new_capacity) {

        do {
            new_content = realloc(q.content, sizeof(T[new_capacity]));
        } while (!new_content);

        if (old_capacity < old_head + length) {
    // fstLength represents the first part of the q
    // that is in the bottom of the array
            size_t const fst_length = old_capacity - old_head;
    // lstLength represents the last part of the q
    // that is in the top of the array
            size_t const lst_length = length - fst_length;
            if (lst_length <= (new_capacity - old_capacity)) {
    // copy the values from the top of the array in to the bottom
                memcpy(
                    new_content + old_capacity,
                    new_content,
                    lst_length * sizeof(unsigned)
                );
            } else {
    // copy the values from the bottom of the array in to the top
                new_head = new_capacity - fst_length;
                memmove(
                    new_content + new_head,
                    new_content + old_head,
                    fst_length * sizeof(unsigned)
                );
            } 
        }
    // never can be newCapacity === oldCapacity
    } else {
        size_t fst_length = length;
        if (old_capacity < old_head + length) {
    // we must copy the bottom values at the end of the new array.
    // fstLength represent the first part of the q
    //that is in the bottom of the array
            fst_length = old_capacity - old_head;
            new_head = new_capacity - fst_length;
        } else {
            new_head = 0;
        }
        memmove(
            q.content + new_head,
            q.content + old_head,
            fst_length * sizeof(unsigned)
        );
         do {
            new_content = realloc(q.content, sizeof(T[new_capacity]));
        } while (!new_content);
    }
    q.content = new_content;
    q.capacity = new_capacity;
    q.head = new_head;
    return q;
}

void queue_free(queue pq[static 1]) {
    free(pq->content);
    *pq = (queue){0};
}

File: JavaScript/array/queue.js
/**
 * Author: Axel Ariel Saravia
 */

/**
Queue T :: {
    capacity: number [uint > 0]
    content: Array T,
    head: number [uint],
    length: number [uint]
}
*/

const ArrayQueue = Object.freeze({
    /**
    create T :: () -> Queue T */
    create() {
        return Object.seal({
            capacity: 1,
            content: Array(1),
            head: 0,
            length: 0
        });
    },
    /**
    resize T :: (Queue T) -> Queue T */
    resize(queue) {
        const length = queue.length;
        const newCapacity = Math.max(1, 2 * length);
    // this is never true when resize is called from add or remove methods
        if (newCapacity < length) {
            return queue;
        }
        const oldCapacity = queue.capacity;
        let oldHead = queue.head;
        let newHead = oldHead;
        if (oldCapacity < newCapacity) {
    // reallocate
            queue.content.length = newCapacity;
            if (oldCapacity < oldHead + length) {
    // fstLength represents the first part of the queue
    // that is in the bottom of the array
                const fstLength = oldCapacity - oldHead;
    // lstLength represents the last part of the queue
    // that is in the top of the array
                const lstLength = length - fstLength;
                if (lstLength <= (newCapacity - oldCapacity)) {
    // copy the values from the top of the array in to the bottom
                    for (let i = 0; i < lstLength; i += 1) {
                        queue.content[i + oldCapacity] = queue.content[i];
                    }
                } else {
    // copy the values from the bottom of the array in to the top
                    newHead = newCapacity - fstLength;
                    for (let i = 0; i < fstLength; i += 1) {
                        queue.content[i + newHead] = queue.content[i + oldHead];
                    }
                }
            }
    // never can be newCapacity === oldCapacity
        } else {
            let fstLength = length;
            if (oldCapacity < oldHead + length) {
    // we must copy the bottom values at the end of the new array.
    // fstLength represent the first part of the queue
    //that is in the bottom of the array
                fstLength = oldCapacity - oldHead;
                newHead = newCapacity - fstLength;
            } else {
                newHead = 0
            }
            for (let i = 0; i < fstLength; i += 1) {
                queue.content[i + newHead] = queue.content[i + oldHead];
            }
    // reallocate
            queue.content.length = newCapacity;
        }
        queue.head = newHead;
        queue.capacity = newCapacity;
        return queue;
    },
    /**
    size T :: (Queue T) -> number [uint] */
    size(queue) {
        return queue.length;
    },
    /**
    get T :: (Queue T) -> T */
    get(queue) {
        return queue.content[queue.head];
    },
    /**
    add T :: (Queue T, T) -> Queue<T */
    add(queue, x) {
        if (queue.length + 1 > queue.capacity) {
            ArrayQueue.resize(queue);
        }
        const idx = (queue.head + queue.length) % queue.capacity;
        queue.content[idx] =  x;
        queue.length += 1;
        return queue;
    },
    /**
    remove T :: (Queue T) -> T */
    remove(queue) {
        const x = queue.content[queue.head];
        queue.head = (queue.head + 1) % queue.capacity;
        queue.length -= 1;
        if (3 * queue.length <= queue.capacity) {
            ArrayQueue.resize(queue);
        }
        return x;
    },
    /**
    clear T :: (Queue T) -> undefined */
    clear(queue) {
        queue.content = Array(1);
        queue.capacity = 1;
        queue.head = 0;
        queue.length = 0;
    }
});

export default ArrayQueue;